include/ginkgo/core/matrix/batch_identity.hpp

```diff
@@ -0,0 +1,210 @@
+            unbatch_mats.emplace_back(std::move(mat));
+        }
+        return unbatch_mats;
+    }
+
+private:
+    /**
+     * Extract sizes from the vector of the distinct Identity matrices.
+     */
+    batch_dim<2> get_sizes_from_mtxs(
+        const std::vector<Identity<ValueType>*> mtxs) const
+    {
+        auto sizes = std::vector<dim<2>>(mtxs.size());
+        for (auto i = 0; i < mtxs.size(); ++i) {
+            sizes[i] = mtxs[i]->get_size();
+        }
+        return batch_dim<2>(sizes);
+    }
+
+protected:
+    /**
+     * Creates an uninitialized BatchIdentity matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the batch matrices in a batch_dim object
+     */
+    BatchIdentity(std::shared_ptr<const Executor> exec,
+                  const batch_dim<2>& size = batch_dim<2>{})
+        : EnableBatchLinOp<BatchIdentity>(exec, size)
+    {}
+
+    /**
+     * Creates a BatchIdentity matrix from a vector of matrices
+     *
+     * @param exec  Executor associated to the matrix
+     * @param matrices  The matrices that need to be batched.
+     */
+    BatchIdentity(std::shared_ptr<const Executor> exec,
+                  const std::vector<Identity<ValueType>*>& matrices)
+        : EnableBatchLinOp<BatchIdentity>(exec, get_sizes_from_mtxs(matrices))
+    {}
+
+    /**
+     * Creates a BatchIdentity matrix by duplicating BatchIdentity matrix
+     *
+     * @param exec  Executor associated to the matrix
+     * @param num_duplications  The number of times to duplicate
+     * @param input  The matrix to be duplicated.
+     */
+    BatchIdentity(std::shared_ptr<const Executor> exec,
+                  const size_type num_duplications,
+                  const BatchIdentity<value_type>* const input)
+        : EnableBatchLinOp<BatchIdentity>(exec)
+    {
+        const auto nbatch1 = input->get_num_batch_entries();
+        if (input->get_size().stores_equal_sizes()) {
+            this->set_size(batch_dim<2>(nbatch1 * num_duplications,
+                                        input->get_size().at(0)));
+        } else {
+            std::vector<dim<2>> sizes(nbatch1 * num_duplications);
+            for (size_type idup = 0; idup < num_duplications; idup++) {
+                for (size_type i = 0; i < nbatch1; i++) {
+                    sizes[idup * nbatch1 + i] = input->get_size().at(i);
+                }
+            }
+            this->set_size(sizes);
+        }
+    }
+
+    /**
+     * Creates a BatchIdentity matrix by duplicating Identity matrix
+     *
+     * @param exec  Executor associated to the matrix
+     * @param num_duplications  The number of times to duplicate
+     * @param input  The matrix to be duplicated.
+     */
+    BatchIdentity(std::shared_ptr<const Executor> exec,
+                  const size_type num_duplications,
+                  const Identity<value_type>* const input)
```

Comment:
Do we really need to take an Identity matrix ? Just a size or a batch_size should be enough right ? Otherwise, the use has to go through the trouble of creating an Identity matrix first.

Reply:
That can already be done by the first constructor. This is there just in case the user already has an identity matrix that they want to duplicate. I guess this can be removed.